django/test/selenium.py

```diff
@@ -83,7 +83,7 @@ def create_options(self):
         options = self.import_options(self.browser)()
         if self.headless:
             match self.browser:
-                case "chrome":
+                case "chrome" | "edge":
```

Comment:
Ran the selenium tests using edge and `admin_inlines.tests.SeleniumTests.test_inline_formset_error_input_border` fails for me. I think this test is generally a bit odd (making asserts on css is not really how we want to test imo) and maybe we update that test to take a screenshot instead?

Reply:
We also have in the docs:
> Some browsers (e.g. Chrome or Firefox) support headless testing

Should we add edge in here?

Reply:
Also would adding supporting for "chromium" be almost the same as these changes?

Reply:
> ...fails for me.

Admittedly I'd only tested the screenshots for Edge. Will have a look later at the wider tests later.

> We also have in the docs: ...

Yes, I noticed, but hadn't put anything in there as it didn't seem to be an exhaustive list (`e.g.`). We could be more explicit, especially given there being a couple of browser-specific options in this file.

> Also would adding supporting for "chromium" be almost the same as these changes?

Should be, yes, they're basically the same. Ditto with Opera. Unfortunately, looking at the [available images](https://hub.docker.com/u/selenium) for Selenium Grid, there isn't one for plain Chromium and the Opera one is deprecated.

Reply:
Looking at the [docs for our selenium unit tests](https://docs.djangoproject.com/en/5.0/internals/contributing/writing-code/unit-tests/#running-the-selenium-tests) it says we support [selenium.webdriver](https://github.com/SeleniumHQ/selenium/tree/trunk/py/selenium/webdriver) package browsers which includes chromium but not Opera.

I didn't even know we used Selenium Grid, but turns out `--selenium-hub` and `--external-host` are undocumented and being used in django-docker-box. [django-docker-box](https://docs.djangoproject.com/en/5.0/internals/contributing/writing-code/unit-tests/#running-tests-using-django-docker-box) _is_ documented but doesn't mention the selenium tests and I would say is "a little bit" untrue that it supports everything (Oracle certainly doesn't work and a number of things are broken there).

We are only running the selenium tests for chrome on PRs and in the daily runs. If we really want to support multiple browsers for a better contributor experience we should probably also pick up [this PR](https://github.com/django/django/pull/17562) and add each browser to part of the daily test run. We may want to explicitly document which browsers we actually support with our selenium tests.

I don't know the best way forward, I would say it is a little bit messy currently 🤔

Reply:
> ...but not Opera.

Oh, I wouldn't worry about Opera (even though it's Chromium-based) - it was an offhand comment based on the image on Docker Hub, but that is 3 years old now that I look at it! 😂 

> (Oracle certainly doesn't work and a number of things are broken there).

It does if you use the [main branch of my fork](https://github.com/ngnpope/django-docker-box/commits/main/) 😉 I've also fixed and improved lots of other things, but it's a work in progress. When I've finished we can work out how to roll it back in.

> django-docker-box is documented but doesn't mention the selenium tests...

And this is also something that I've fixed up to work correctly in my fork 💪🏻 (Proper documentation of everything is still to come.)

> We are only running the selenium tests for chrome on PRs and in the daily runs.

I see. I wasn't aware that Firefox was also broken. Do you have any objection to the commit that allows Edge to at least run correctly in headless mode and also to do the high constrast screenshots (which works for me)? ISTM that, maybe when I have a workflow that can "run the world" in my fork, we'll be able to more reliably test and fix Edge/Firefox later.

Reply:
> It does if you use the [main branch of my fork](https://github.com/ngnpope/django-docker-box/commits/main/) 😉

Amazing. If that's true that's so so good

> Do you have any objection to the commit that allows Edge to at least run correctly in headless mode and also to do the high constrast screenshots (which works for me)?

No objections 👍

Reply:
> Amazing. If that's true that's so so good

😆 It's off topic, but you gotta see this:

```console
$ docker compose ps
NAME                                    IMAGE                                                  COMMAND                  SERVICE             CREATED          STATUS                    PORTS
django-docker-box-mariadb-db-1          mariadb:10.5                                           "docker-entrypoint.s…"   mariadb-db          5 minutes ago    Up 5 minutes (healthy)    3306/tcp
django-docker-box-mariadb-gis-db-1      mariadb:10.5                                           "docker-entrypoint.s…"   mariadb-gis-db      3 minutes ago    Up 3 minutes (healthy)    3306/tcp
django-docker-box-memcached-1-1         memcached:alpine                                       "docker-entrypoint.s…"   memcached-1         19 minutes ago   Up 18 minutes (healthy)   11211/tcp
django-docker-box-memcached-2-1         memcached:alpine                                       "docker-entrypoint.s…"   memcached-2         18 minutes ago   Up 18 minutes (healthy)   11211/tcp
django-docker-box-mysql-db-1            mysql:8                                                "docker-entrypoint.s…"   mysql-db            5 minutes ago    Up 5 minutes (healthy)    3306/tcp, 33060/tcp
django-docker-box-mysql-gis-db-1        mysql:8                                                "docker-entrypoint.s…"   mysql-gis-db        3 minutes ago    Up 3 minutes (healthy)    3306/tcp, 33060/tcp
django-docker-box-oracle-db-1           container-registry.oracle.com/database/free:23.3.0.0   "/bin/bash -c $ORACL…"   oracle-db           5 minutes ago    Up 5 minutes (healthy)    1521/tcp
django-docker-box-oracle-gis-db-1       container-registry.oracle.com/database/free:23.3.0.0   "/bin/bash -c $ORACL…"   oracle-gis-db       3 minutes ago    Up 3 minutes (healthy)    1521/tcp
django-docker-box-postgresql-db-1       postgres:13-alpine                                     "docker-entrypoint.s…"   postgresql-db       5 minutes ago    Up 5 minutes (healthy)    5432/tcp
django-docker-box-postgresql-gis-db-1   postgis/postgis:13-3.0                                 "docker-entrypoint.s…"   postgresql-gis-db   3 minutes ago    Up 3 minutes (healthy)    5432/tcp
django-docker-box-redis-1               redis:alpine                                           "docker-entrypoint.s…"   redis               18 minutes ago   Up 18 minutes (healthy)   6379/tcp
django-docker-box-selenium-chrome-1     selenium/node-chrome                                   "/opt/bin/entry_poin…"   selenium-chrome     24 hours ago     Up 24 hours               5900/tcp
django-docker-box-selenium-edge-1       selenium/node-edge                                     "/opt/bin/entry_poin…"   selenium-edge       23 hours ago     Up 23 hours               5900/tcp
django-docker-box-selenium-firefox-1    selenium/node-firefox                                  "/opt/bin/entry_poin…"   selenium-firefox    10 minutes ago   Up 10 minutes             5900/tcp
django-docker-box-selenium-hub-1        selenium/hub                                           "/opt/bin/entry_poin…"   selenium-hub        24 hours ago     Up 24 hours (healthy)     0.0.0.0:4442-4444->4442-4444/tcp, :::4442-4444->4442-4444/tcp
```